specs/merkle-tree.md

```diff
@@ -0,0 +1,158 @@
+
+### Input
+Accepts an object where the keys are canonical chain ids and the values are deployment data objects which contain all of the necessary info to assemble a SphinxMerkleTree. See the [inline documentation](TODO(end)) for exact details on the input data. The Sphinx Merkle tree generation function should be agnostic to the source of the transaction data so it can be used to assemble a tree based on transactions generated from any scripting framework (i.e Foundry, Hardhat Ignition, some arbitrary future framework).
+
+### Output
+A Sphinx Merkle tree that follows the above specification and invariants. See the [inline documetation](TODO(end)) for the exact details of the output data structure.
+
+## Function-Level Invariants
+We export three separate functions `makeSphinxLeaves`, `makeSphinxMerkleTreeFromLeaves`, and `makeSphinxMerkleTree`. We provide three separate functions for convenience and modularity to enable testing. However, for this specification and the below invariants we assume the use of the `makeSphinxMerkleTree`. We *only* recommend using that function when generating Sphinx Merkle trees in production and therefore only consider it to satisfy the invariants listed above.
+
+### It is impossible to specify multiple `NetworkDeploymentData` objects for a given network
+The Merkle tree generation logic relies on the fact that it is not possible to submit multiple `NetworkDeploymentData` objects to guarantee that there is exactly one `APPROVE` or `CANCEL` leaf ([4](#4-there-must-be-exactly-one-approval-leaf-or-cancel-leaf-per-chain-4)).
+
+### The index of all leaves is generated internally and cannot be specified by the input data
+The Merkle tree generation logic relies on calculating the indexes for all leaves internally to guarantee that the leaves are ordered properly ([2](#2-merkle-tree-leaves-must-be-ordered-in-the-tree-by-index-and-chain-id-ascending)) and that there are no index gaps ([1](#1-there-must-be-no-gaps-in-index-in-leaves-that-have-the-same-chain-id)).
+
+This invariant works together with the above invariant, it is not possible to specify multiple `NetworkDeploymentData` objects for a given network, to ensure each index + chain ID combination is unique ([3](#3-every-chain-id-and-index-combination-must-be-unique-within-the-tree-3)).
+
+## Dependencies
+The Sphinx standard Merkle tree generation logic makes calls to two popular external libraries. We use `ethers` version 6.7.0 to handle ABI encoding tree leaf `data`, and we use `@openzeppelin/merkle-tree` to assemble the Merkle tree. We test that interactions with these libraries work properly in our higher-level unit tests which ensure that Merkle trees generated by this logic have data that is encoded properly and that they are executable on-chain. However, we do not test the internals of these libraries and instead rely on the assumption that they are robust and bug-free.
```

Comment:
> our higher-level unit tests which ensure that... they are executable on-chain.

It seems like you're referring to the Merkle tree unit tests, which don't check that the tree is executable on-chain. I.e. the `SphinxModuleProxy` can revert for any number of reasons, which isn't checked by the Merkle tree tests

Reply:
I'm referring to `SphinxModuleProxy.t.sol` which uses this function to generate Merkle trees and tests out executing them.